nakama/events.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/heroiclabs/nakama-common/runtime"
+)
+
+type Event struct {
+	message string
+}
+
+type connectionChannelPairs struct {
+	connection *websocket.Conn
+	channel    chan *Event
+}
+
+type EventHub struct {
+	inputConnection *websocket.Conn
+	channels        *sync.Map //map[string]chan *Event
+	didShutdown     atomic.Bool
+}
+
+func createEventHub() (*EventHub, error) {
+	url := makeWebSocketURL(eventEndpoint)
+	webSocketConnection, _, err := websocket.DefaultDialer.Dial(url, nil)
+	for err != nil {
+		if errors.Is(err, &net.DNSError{}) {
+			//sleep a little try again...
+			fmt.Println("No host found.")
+			fmt.Println(err.Error())
+			time.Sleep(2 * time.Second)
+			webSocketConnection, _, err = websocket.DefaultDialer.Dial(url, nil)
+		} else {
+			return nil, err
+		}
+	}
+	channelMap := sync.Map{}
+	res := EventHub{
+		inputConnection: webSocketConnection,
+		channels:        &channelMap,
+		didShutdown:     atomic.Bool{},
+	}
+	res.didShutdown.Store(false)
+	return &res, nil
+}
+
+func (eh *EventHub) subscribe(session string) chan *Event {
+	channel := make(chan *Event)
+	eh.channels.Store(session, channel)
+	return channel
+}
+
+func (eh *EventHub) unsubscribe(session string) {
```

Comment:
Since EventHub is the sender on this channel, it should also close it before deleting it from the map.